Separate limits scaling between CPU & memory

sibucan opened the pull request:
See the [comment](https://github.com/kubernetes/autoscaler/issues/3965#issuecomment-851024267) I made for issue #3965

When the ratio between the original memory requests/limits on the Deployment is a floating point number instead of an integer, the calculated limit is expressed in millivalues, (https://play.golang.org/p/dR4SaXRBGmB) which is valid for CPU values, but not necessarily for memory values.

The purpose of this PR is to ensure it is expressed as bytes.

jbartosik commented on vertical-pod-autoscaler/pkg/utils/vpa/limit_and_request_scaling.go:
```diff
@@ -107,7 +115,25 @@ func GetBoundaryRequest(originalRequest, originalLimit, boundaryLimit, defaultLi
+		return &resource.Quantity{}
+	}
+	// originalLimit set but originalRequest not set - K8s will treat the pod as if they were equal
+	if originalRequest == nil || originalRequest.Value() == 0 {
```
Why this is not in `GetBoundaryRequestCPU`?  I think it should be there.

sibucan replied:
It should be there, since `GetBoundaryRequestCPU` is just the original function, but renamed. Nothing was removed from it, rather a copy of it was created and named `GetBoundaryRequestMem` to get the scaled memory values (as opposed to scaled CPU values).

jbartosik commented on vertical-pod-autoscaler/pkg/utils/vpa/limit_and_request_scaling.go:
```diff
@@ -107,7 +115,25 @@ func GetBoundaryRequest(originalRequest, originalLimit, boundaryLimit, defaultLi
+
+// GetBoundaryRequestMem returns the boundary (min/max) memory request that can be specified with
+// preserving the original limit to request ratio. Returns nil if no boundary exists
+func GetBoundaryRequestMem(originalRequest, originalLimit, boundaryLimit, defaultLimit *resource.Quantity) *resource.Quantity {
```
`GetBoundaryRequestMem` is (and should be) very similar to `GetBoundaryRequestCPU` (it looks like the only difference is that CPU operates on milli units and Memory operates on whole values). I'd rather avoid having two very similar pieces of code

sibucan replied:
That's fair -- I could revert these changes and keep the original function name, and instead add a new input parameter to the function (e.g. "cpu" or "memory") which will determine what scaling function to use. How does that sound?

jbartosik replied:
Sounds good to me

jbartosik reviewed (changes requested):
Thanks & sorry for slow reply.

Please add some tests that cover cases when we expect this to behave differently

sibucan commented:
@jbartosik made changes based on feedback & added tests for the changes made in the scaling function, lmk if you find any other issues

jbartosik reviewed (changes requested):
A couple more small changes

sibucan commented on vertical-pod-autoscaler/pkg/utils/vpa/limit_and_request_scaling.go:
```diff
@@ -141,3 +156,26 @@ func scaleQuantityProportionally(scaledQuantity, scaleBase, scaleResult *resourc
 }
+
+// scaleQuantityProportionallyMem returns value which has the same proportion to scaledQuantity as scaleResult has to scaleBase
+// It also returns a bool indicating if it had to cap result to MaxInt64 milliunits.
```
Good catch, updated.

sibucan commented on vertical-pod-autoscaler/pkg/utils/vpa/limit_and_request_scaling.go:
```diff
@@ -119,9 +134,9 @@ const (
 )
 
-// scaleQuantityProportionally returns value which has the same proportion to scaledQuantity as scaleResult has to scaleBase
+// scaleQuantityProportionallyCPU returns value which has the same proportion to scaledQuantity as scaleResult has to scaleBase
```
Mentioned how the value returned can be either in units or in milliunits depending on the function being called.

jbartosik reviewed (commented):
Thank you. Looks good to me now. Please squash then I'll approve this PR

sibucan commented:
Commits have been squashed together to a single one.

sibucan commented:
@jbartosik I noticed that some tests were failing -- one of the tests failures was a simple change due to how memory resources are now displayed correctly, but the other failure was a bit more complicated to debug and it's related to the [rounding](https://github.com/kubernetes/autoscaler/blob/68a7d27b84d803eceadb2d5e82aff4dc517318d2/vertical-pod-autoscaler/pkg/utils/vpa/limit_and_request_scaling.go#L133) logic.

It seems that previously, the rounding for memory values worked by taking advantage of the millivalue representation:
```
result: 357913941333m
round down: 357913941
result: 715827882666m
round down: 715827882

result: 1431655765333m
round up: 1431655766
result: 2863311530666m
round up: 2863311531
```

However, after these changes, the results just don't round the same way anymore:
```
result: 357913941
round down: 357913941
result: 715827882
round down: 715827882

result: 1431655765
round up: 1431655765
result: 2863311530
round up: 2863311530
```

This causes some of the tests to fail. I'm not sure how to fix this, since changing the [rounding scale](https://github.com/kubernetes/autoscaler/blob/68a7d27b84d803eceadb2d5e82aff4dc517318d2/vertical-pod-autoscaler/pkg/utils/vpa/limit_and_request_scaling.go#L134) only makes the rounded result differ from current tests even more, e.g. using a [scale of 1](https://github.com/kubernetes/apimachinery/blob/a644435e2c133d990b624858d9c01985d7f59adf/pkg/api/resource/quantity_test.go#L516) would make `1431655765` round up to `1431655770`, which is technically correct, but the tests expect `1431655766`.

I changed the expected values to reflect the new actuals, but I'm not sure of the implications of doing this. Any advice would be appreciated.

jbartosik commented:
@sibucan I still see two commits. Please squash

sibucan commented:
@jbartosik Hi! I just thought you'd want to re-review that other commit by itself before I squashed commits, but I went ahead and squashed again in the interest of time.

jbartosik reviewed (approved):
/lgtm
/approve
